esmvaltool/diag_scripts/landcover/albedolandcover.py

```diff
@@ -0,0 +1,338 @@
+            assert(model_data['snc'].data[i, j] < dia_cfg['thres_fsnow'])
+            # Now check the local neighbourhood.
+            # Create indices
+            # Create the neighbourhood as bbox
+            islice = slice(int(i - (dia_cfg['lonsize_BB'] - 1) / 2),
+                           int(i + (dia_cfg['lonsize_BB'] - 1) / 2 + 1))
+            jslice = slice(int(j - (dia_cfg['latsize_BB'] - 1) / 2),
+                           int(j + (dia_cfg['latsize_BB'] - 1) / 2 + 1))
+            bbox_mask = model_data['alb'].data.mask[islice, jslice]
+
+            # Check if there are enough valid data points
+            # in the neighbourhood bbox
+            if np.sum((~bbox_mask).astype(int)) > dia_cfg['minnum_gc_bb']:
+                lc_logical = np.full((3, ), True)
+                lc_classes = [dia_cfg['lc1_class'], dia_cfg['lc2_class'],
+                              dia_cfg['lc3_class']]
+                lc_data = []
+                # Loop over lc_classes
+                for i_0 in range(3):
+                    current_class = lc_classes[i_0]
+                    # First flatten the array
+                    lc_flattened = {}
+                    for varkey in current_class:
+                        lc_flattened[varkey] = model_data[varkey].data[
+                            islice, jslice].compressed()
+                    lc_sum = sum([lc_flattened[varkey]
+                                  for varkey in current_class])
+                    # Now check thresholds
+                    if (np.var(lc_sum) > 0. and
+                            len(lc_sum) >= dia_cfg['mingc']):
+                        lc_data.append(lc_sum)
+                        lc_logical[i_0] = True
+                    else:
+                        print("Variance zero or not enough\
+                               valid data for this landcover class")
+                        lc_logical[i_0] = False
+                # Now the multiple lin reg part
+                x_0 = np.stack(lc_data)
+                x_0 = x_0.swapaxes(0, 1)
+                # Same mask, so shape is fine
+                y_0 = model_data['alb'].data[islice, jslice].compressed()
+
+                # Check that the system is not over_parameterised
+                if len(y_0) > np.sum(lc_logical.astype(int)) + 1 and\
+                        np.sum(lc_logical.astype(int)) > 0:
+                    # Do multiple linear regression
+                    linreg = linear_model.LinearRegression().fit(x_0, y_0)
+                    intercept = linreg.intercept_
+                    coefficients = linreg.coef_
+
+                    # Now loop again and reconstruct albedo's
+                    lc_reg = 0
+                    for i_0 in range(3):
+                        if lc_logical[i_0]:
+                            alb_lc[i_0, i, j] = intercept\
+                                + coefficients[lc_reg] * 100.
+                            lc_reg = lc_reg + 1
+    return alb_lc
```

Comment:
this is a *very* complex nested loop, honestly I don't understand the science behind it so it might be there is no way to simplify it, but is there a way to simplify it/try reduce the number of branches?